src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStoreEnumerator.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
+        {
+            _values = value;
+            _value = null;
+        }
+
+        public Enumerator GetEnumerator() => _values != null ? new Enumerator(_values!) : new Enumerator(_value!);
+
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<string>
+        {
+            private readonly IEnumerator<string>? _values;
+            private readonly string? _value;
+            private bool _completed;
+
+            internal Enumerator(IEnumerable<string> values)
+            {
+                _values = values.GetEnumerator();
+                _value = null;
+                _completed = false;
+                Current = default!;
+            }
+
+            internal Enumerator(string value)
+            {
+                _values = null;
+                _value = value;
+                _completed = false;
+                Current = default!;
+            }
+
+            public bool MoveNext()
+            {
+                if (_completed)
+                {
+                    return false;
+                }
+
+                if (_values == null)
+                {
+                    _completed = true;
+                    Current = _value!;
+                    return true;
+                }
+
+                _completed = !_values.MoveNext();
+                if (_completed)
+                {
+                    return false;
+                }
+
+                Debug.Assert(_values.Current != null);
+                Current = _values.Current!;
+
+                return true;
+            }
+
+            public string Current { get; private set; }
```

Comment:
You have this defined as `readonly` in the ref.  The get here needs to be made `readonly`.